Source/Core/Core/HW/WiimoteEmu/MotionPlus.cpp

```diff
@@ -340,26 +489,61 @@ void MotionPlus::Update()
+      WARN_LOG(WIIMOTE, "M+ unknown passthrough-mode %d", (int)GetPassthroughMode());
+      mplus_data.is_mp_data = true;
       break;
     }
   }
 
   // If the above logic determined this should be M+ data, update it here
   if (mplus_data.is_mp_data)
   {
-    // Wiibrew: "While the Wiimote is still, the values will be about 0x1F7F (8,063)"
-    // high-velocity range should be about +/- 1500 or 1600 dps
-    // low-velocity range should be about +/- 400 dps
-    // Wiibrew implies it shoould be +/- 595 and 2700
+    // These are the max referene velocities used by the sensor of the M+.
+    // TODO: Reverse engineer the calibration data to send perfect values.
+    constexpr float SLOW_MAX_RAD_PER_SEC = 440 * float(MathUtil::TAU) / 360;
+    constexpr float FAST_MAX_RAD_PER_SEC = 2000 * float(MathUtil::TAU) / 360;
+
+    constexpr int BITS_OF_PRECISION = 14;
+    constexpr s32 MAX_VALUE = (1 << BITS_OF_PRECISION) - 1;
+
+    // constexpr u16 NEUTRAL_YAW = 0x1f66;
+    // constexpr u16 NEUTRAL_ROLL = 0x2058;
+    // constexpr u16 NEUTRAL_PITCH = 0x1fa8;
+
+    constexpr u16 NEUTRAL_YAW = 0x1f2e;
+    constexpr u16 NEUTRAL_ROLL = 0x1f72;
+    constexpr u16 NEUTRAL_PITCH = 0x1f9d;
+
+    // constexpr u16 SENSOR_NEUTRAL = (1 << (BITS_OF_PRECISION - 1));
+    // constexpr u16 SENSOR_NEUTRAL = 0x783a >> 2;
+    constexpr u16 SENSOR_RANGE = (1 << (BITS_OF_PRECISION - 1));
+
+    constexpr float SLOW_SCALE = SENSOR_RANGE / SLOW_MAX_RAD_PER_SEC;
+    constexpr float FAST_SCALE = SENSOR_RANGE / FAST_MAX_RAD_PER_SEC;
+
+    const float yaw = angular_velocity.z;
+    // TODO: verify roll signedness with our calibration data.
+    const float roll = angular_velocity.y;
+    const float pitch = angular_velocity.x;
+
+    // Slow scaling can be used if it fits in the sensor range.
+    mplus_data.yaw_slow = (std::abs(yaw) < SLOW_MAX_RAD_PER_SEC);
+    s32 yaw_value = yaw * (mplus_data.yaw_slow ? SLOW_SCALE : FAST_SCALE);
+
+    mplus_data.roll_slow = (std::abs(roll) < SLOW_MAX_RAD_PER_SEC);
+    s32 roll_value = roll * (mplus_data.roll_slow ? SLOW_SCALE : FAST_SCALE);
+
+    mplus_data.pitch_slow = (std::abs(pitch) < SLOW_MAX_RAD_PER_SEC);
+    s32 pitch_value = pitch * (mplus_data.pitch_slow ? SLOW_SCALE : FAST_SCALE);
 
-    u16 yaw_value = 0x2000;
-    u16 roll_value = 0x2000;
-    u16 pitch_value = 0x2000;
+    yaw_value = MathUtil::Clamp(yaw_value + NEUTRAL_YAW, 0, MAX_VALUE);
+    roll_value = MathUtil::Clamp(roll_value + NEUTRAL_ROLL, 0, MAX_VALUE);
+    pitch_value = MathUtil::Clamp(pitch_value + NEUTRAL_PITCH, 0, MAX_VALUE);
 
-    mplus_data.yaw_slow = 1;
-    mplus_data.roll_slow = 1;
-    mplus_data.pitch_slow = 1;
+    // INFO_LOG(WIIMOTE, "M+ YAW: 0x%x slow:%d", yaw_value, mplus_data.yaw_slow);
+    // INFO_LOG(WIIMOTE, "M+ ROL: 0x%x slow:%d", roll_value, mplus_data.roll_slow);
+    // INFO_LOG(WIIMOTE, "M+ PIT: 0x%x slow:%d", pitch_value, mplus_data.pitch_slow);
```

Comment:
I feel those would be spammy if we left them in; I _think_ we should remove those completely before merging.

Reply:
I'd love it if we had log configuration like Log4J where you can set different verbosity levels for different components. That would allow us to have easily accessible debugging logs for something we need to diagnose without spamming the (visible) log with unrelated data …